### ISSUE TITLE
Supporting Lambda by using multiprocessing.Pipe

### ISSUE THREAD

**Issue by konklone** [nabla-c0d3/sslyze#252]
I'm working on a Lambda-based SSLyze pipeline, and one thing I came across is that Lambda doesn't support `multiprocessing.Queue` or `multiprocessing.Pool`:

https://aws.amazon.com/blogs/compute/parallel-processing-in-python-with-aws-lambda/

But it does support `multiprocessing.Pipe`. The [multiprocessing docs](https://docs.python.org/3/library/multiprocessing.html) say:

> The Pipe() function returns a pair of connection objects connected by a pipe which by default is duplex (two-way). For example:

```python
from multiprocessing import Process, Pipe

def f(conn):
    conn.send([42, None, 'hello'])
    conn.close()

if __name__ == '__main__':
    parent_conn, child_conn = Pipe()
    p = Process(target=f, args=(child_conn,))
    p.start()
    print(parent_conn.recv())   # prints "[42, None, 'hello']"
    p.join()
```
> The two connection objects returned by Pipe() represent the two ends of the pipe. Each connection object has send() and recv() methods (among others). Note that data in a pipe may become corrupted if two processes (or threads) try to read from or write to the same end of the pipe at the same time. Of course there is no risk of corruption from processes using different ends of the pipe at the same time.

Would it be tricky or dangerous to use `multiprocessing.Pipe` inside of SSLyze's concurrent scanning system, instead of `multiprocessing.Queue`? Are there any considerations or constraints it would impose on users?

**Comment by nabla-c0d3** [comment]
SSLyze on Lambda sounds pretty cool. 
From looking at https://stackoverflow.com/questions/8463008/python-multiprocessing-pipe-vs-queue , it seems like the concurrent scanner would be need to be drastically changed as a `Pipe` does not support multiple consumers? We use the `Queue` specifically to dispatch scan tasks across subprocesses.

**Comment by konklone** [comment]
That's true, it sounds like you would need to manage multiple `Pipe`s yourself (basically reimplementing a `Queue`). 

I haven't looked at the source code for `Queue`, but based on Amazon's post, it looks like `Queue` is implemented using `/dev/shm` to share memory between the generated processes. So I guess one way to do that would be to either reimplement or fork a version of `multiprocessing.Queue` that doesn't use shared memory.

It's not the end of the world to not use a concurrent scanner in a Lambda environment, and potentially if I cared about it enough, I could splinter out the Lambda processes so I was running a Lambda container per-scanner rather than per-hostname. In practice, for my own purposes (where I run 6 scanners), I'm likely to just eat the time cost of using the synchronous scanner instead, but I could see this being a bigger deal for use cases involving more scanners.

**Comment by nabla-c0d3** [comment]
Sounds good - I would have done it if it was an easy change, but it's not the case and would be too AWS-centric for this to be part of SSLyze.

**Comment by konklone** [comment]
Yep, understood. For reference for the thread, I ended up implementing a Lambda pipeline:
https://github.com/18F/domain-scan/blob/master/docs/lambda.md#using-domain-scan-with-amazon-lambda

Which uses SSLyze with a SynchronousScanner:
https://github.com/18F/domain-scan/blob/master/scanners/sslyze.py

Running many SSLyze scans at once using a pool of hundreds simultaneous Lambda functions is quite fast, and reasonably reliable. I ran ~22,000 SSLyze scans using a rolling pool of 800 Lambda functions in about 9 and a half minutes (~38 scans / second).

I still have a small set of scans which will get killed for running out of memory, but from what I've seen so far, I think this is related to a specific set of pathological cases that balloon the memory, and it ca generally be run with 128MB or 256MB containers.
